File: apps/api/momento_api/processor/importer.py
import posixpath
import threading
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from pathlib import Path, PurePosixPath
from typing import Optional

from webdav3.client import Client

from momento_api.config import ReverseGeocodingConfig
from momento_api.constants import IMPORTS_DIR, SUPPORTED_EXTENSIONS
from momento_api.processor.media_processor import process_media_file


class ImportStatus(str, Enum):
    IDLE = "idle"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class ImportJob:
    status: ImportStatus = ImportStatus.IDLE
    total_files: int = 0
    processed_files: int = 0
    successful_imports: int = 0
    failed_imports: int = 0
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    errors: list[str] = field(default_factory=list)


_current_job: Optional[ImportJob] = None
_job_lock = threading.Lock()


def get_import_status() -> ImportJob:
    with _job_lock:
        if _current_job is None:
            return ImportJob()
        return ImportJob(
            status=_current_job.status,
            total_files=_current_job.total_files,
            processed_files=_current_job.processed_files,
            successful_imports=_current_job.successful_imports,
            failed_imports=_current_job.failed_imports,
            started_at=_current_job.started_at,
            completed_at=_current_job.completed_at,
            errors=list(_current_job.errors),
        )


def is_import_running() -> bool:
    with _job_lock:
        return _current_job is not None and _current_job.status == ImportStatus.RUNNING


def _start_import_job() -> ImportJob:
    global _current_job

    with _job_lock:
        if _current_job is not None and _current_job.status == ImportStatus.RUNNING:
            return _current_job
        _current_job = ImportJob(status=ImportStatus.RUNNING, started_at=datetime.now())
        return _current_job


def _finalize_job_success() -> None:
    with _job_lock:
        if _current_job is None:
            return
        _current_job.status = ImportStatus.COMPLETED
        _current_job.completed_at = datetime.now()


def _finalize_job_failure(message: str) -> None:
    with _job_lock:
        if _current_job is None:
            return
        _current_job.status = ImportStatus.FAILED
        _current_job.completed_at = datetime.now()
        _current_job.errors.append(message)


def _update_job_totals(total_files: int) -> None:
    with _job_lock:
        if _current_job is None:
            return
        _current_job.total_files = total_files


def _update_job_progress(success: bool, error_message: Optional[str]) -> None:
    with _job_lock:
        if _current_job is None:
            return
        _current_job.processed_files += 1
        if success:
            _current_job.successful_imports += 1
            return
        _current_job.failed_imports += 1
        if error_message:
            _current_job.errors.append(error_message)


def _collect_import_files(root: Path) -> list[Path]:
    files_to_import: list[Path] = []
    for ext in SUPPORTED_EXTENSIONS:
        files_to_import.extend(root.glob(f"*{ext}"))
        files_to_import.extend(root.glob(f"*{ext.upper()}"))
        files_to_import.extend(root.glob(f"**/*{ext}"))
        files_to_import.extend(root.glob(f"**/*{ext.upper()}"))

    return list(set(files_to_import))


def run_local_import(
    user_id: int,
    thumbnail_max_size: int,
    thumbnail_quality: int,
    video_frame_quality: int,
    delete_after_import: bool,
    reverse_geo_config: Optional[ReverseGeocodingConfig] = None,
) -> None:
    _start_import_job()

    try:
        files_to_import = _collect_import_files(IMPORTS_DIR)
        _update_job_totals(len(files_to_import))

        for file_path in files_to_import:
            if not file_path.exists():
                _update_job_progress(success=False, error_message=f"Missing file: {file_path.name}")
                continue

            media_id = process_media_file(
                file_path, user_id, thumbnail_max_size, thumbnail_quality, video_frame_quality, reverse_geo_config
            )

            if media_id is None:
                _update_job_progress(success=False, error_message=f"Failed to process: {file_path.name}")
                continue

            if delete_after_import:
                try:
                    file_path.unlink(missing_ok=True)
                except OSError as error:
                    _update_job_progress(success=False, error_message=f"Failed to delete {file_path.name}: {error}")
                    continue

            _update_job_progress(success=True, error_message=None)

        _finalize_job_success()

    except (OSError, ValueError) as error:
        _finalize_job_failure(f"Import failed: {error}")


def _ensure_webdav_path(path: str) -> str:
    if not path:
        return "/"
    normalized = posixpath.normpath(path)
    if normalized == ".":
        normalized = "/"
    if not normalized.startswith("/"):
        normalized = f"/{normalized}"
    return normalized


def _list_webdav_files(client: Client, remote_path: str) -> list[str]:
    try:
        return client.list(remote_path)
    except (OSError, ValueError):
        return []


def _collect_webdav_files(client: Client, remote_path: str) -> list[str]:
    stack = [remote_path]
    files: list[str] = []

    while stack:
        current = stack.pop()
        entries = _list_webdav_files(client, current)
        for entry in entries:
            if entry in (".", ".."):
                continue
            full_path = posixpath.join(current, entry)
            if full_path.endswith("/"):
                stack.append(full_path.rstrip("/"))
            else:
                files.append(full_path)

    return files


def _filter_supported_remote_files(files: list[str]) -> list[str]:
    supported: list[str] = []
    for file_path in files:
        ext = PurePosixPath(file_path).suffix.lower()
        if ext in SUPPORTED_EXTENSIONS:
            supported.append(file_path)
    return supported


def run_webdav_import(
    user_id: int,
    hostname: str,
    username: str,
    password: str,
    remote_path: str,
    thumbnail_max_size: int,
    thumbnail_quality: int,
    video_frame_quality: int,
    reverse_geo_config: Optional[ReverseGeocodingConfig] = None,
) -> None:
    _start_import_job()

    try:
        options = {"webdav_hostname": hostname, "webdav_login": username, "webdav_password": password}
        client = Client(options)
        base_path = _ensure_webdav_path(remote_path)
        remote_files = _collect_webdav_files(client, base_path)
        supported_files = _filter_supported_remote_files(remote_files)
        _update_job_totals(len(supported_files))

        for remote_file in supported_files:
            local_target = IMPORTS_DIR / PurePosixPath(remote_file).name
            try:
                client.download_sync(remote_path=remote_file, local_path=str(local_target))
            except OSError as error:
                _update_job_progress(success=False, error_message=f"Failed to download {remote_file}: {error}")
                continue

            media_id = process_media_file(
                local_target, user_id, thumbnail_max_size, thumbnail_quality, video_frame_quality, reverse_geo_config
            )

            if media_id is None:
                _update_job_progress(success=False, error_message=f"Failed to process: {remote_file}")
                continue

            try:
                local_target.unlink(missing_ok=True)
            except OSError as error:
                _update_job_progress(success=False, error_message=f"Failed to delete {remote_file}: {error}")
                continue

            _update_job_progress(success=True, error_message=None)

        _finalize_job_success()

    except (OSError, ValueError) as error:
        _finalize_job_failure(f"WebDAV import failed: {error}")

File: apps/api/src/processor/importer.rs
use chrono::{DateTime, Utc};
use std::fmt;
use std::path::{Path, PathBuf};
use std::sync::RwLock;
use tracing::{debug, error, info, warn};

use crate::config::Config;
use crate::constants::{IMPORTS_DIR, SUPPORTED_EXTENSIONS, WEBDAV_DIR};
use crate::database::{fetch_one, DbPool};
use crate::processor::media_processor::{process_media_file, MediaProcessingContext};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ImportStatus {
    Idle,
    Running,
    Completed,
    Failed,
}

impl fmt::Display for ImportStatus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ImportStatus::Idle => write!(f, "idle"),
            ImportStatus::Running => write!(f, "running"),
            ImportStatus::Completed => write!(f, "completed"),
            ImportStatus::Failed => write!(f, "failed"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct ImportJob {
    pub status: ImportStatus,
    pub total_files: i64,
    pub processed_files: i64,
    pub successful_imports: i64,
    pub failed_imports: i64,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub errors: Vec<String>,
}

#[derive(Clone)]
pub struct ImportSettings {
    pub processing: MediaProcessingContext,
    pub delete_after_import: bool,
    pub concurrency: usize,
}

impl Default for ImportJob {
    fn default() -> Self {
        Self {
            status: ImportStatus::Idle,
            total_files: 0,
            processed_files: 0,
            successful_imports: 0,
            failed_imports: 0,
            started_at: None,
            completed_at: None,
            errors: Vec::new(),
        }
    }
}

/// Maximum number of errors to store in job state to prevent unbounded memory growth
const MAX_JOB_ERRORS: usize = 100;

lazy_static::lazy_static! {
    static ref CURRENT_JOB: RwLock<ImportJob> = RwLock::new(ImportJob::default());
}

fn push_job_error(errors: &mut Vec<String>, message: &str) {
    if errors.len() < MAX_JOB_ERRORS {
        errors.push(message.to_string());
    } else if errors.len() == MAX_JOB_ERRORS {
        errors.push("(additional errors truncated)".to_string());
    }
}

pub fn get_import_status() -> ImportJob {
    CURRENT_JOB.read().unwrap().clone()
}

pub fn is_import_running() -> bool {
    CURRENT_JOB.read().unwrap().status == ImportStatus::Running
}

fn start_import_job() {
    let mut job = CURRENT_JOB.write().unwrap();
    if job.status == ImportStatus::Running {
        return;
    }
    *job = ImportJob {
        status: ImportStatus::Running,
        started_at: Some(Utc::now()),
        ..Default::default()
    };
}

fn finalize_job_success() {
    let mut job = CURRENT_JOB.write().unwrap();
    job.status = ImportStatus::Completed;
    job.completed_at = Some(Utc::now());
}

#[allow(dead_code)]
fn finalize_job_failure(message: &str) {
    let mut job = CURRENT_JOB.write().unwrap();
    job.status = ImportStatus::Failed;
    job.completed_at = Some(Utc::now());
    push_job_error(&mut job.errors, message);
}

fn update_job_totals(total_files: i64) {
    let mut job = CURRENT_JOB.write().unwrap();
    job.total_files = total_files;
}

fn update_job_progress(success: bool, error_message: Option<&str>) {
    let mut job = CURRENT_JOB.write().unwrap();
    job.processed_files += 1;
    if success {
        job.successful_imports += 1;
    } else {
        job.failed_imports += 1;
        if let Some(msg) = error_message {
            push_job_error(&mut job.errors, msg);
        }
    }
}

fn collect_import_files(root: &Path) -> Vec<PathBuf> {
    let mut files = Vec::new();

    for ext in SUPPORTED_EXTENSIONS.iter() {
        // Collect files with both cases
        let patterns = vec![
            format!("**/*{}", ext),
            format!("**/*{}", ext.to_uppercase()),
            format!("*{}", ext),
            format!("*{}", ext.to_uppercase()),
        ];

        for pattern in patterns {
            let glob_pattern = root.join(&pattern);
            if let Ok(paths) = glob::glob(glob_pattern.to_str().unwrap_or("")) {
                for path in paths.filter_map(Result::ok) {
                    if path.is_file() && !files.contains(&path) {
                        files.push(path);
                    }
                }
            }
        }
    }

    files
}

use futures::stream::{self, StreamExt};
use std::sync::Arc;
use tokio::sync::Semaphore;
use tokio::task::JoinSet;

pub async fn run_local_import(settings: ImportSettings) {
    start_import_job();

    let files_to_import = collect_import_files(&IMPORTS_DIR);
    update_job_totals(files_to_import.len() as i64);

    let effective_concurrency = if settings.concurrency > 0 {
        settings.concurrency
    } else {
        num_cpus::get()
    };
    let semaphore = Arc::new(Semaphore::new(effective_concurrency));
    let delete_after_import = settings.delete_after_import;
    let processing = settings.processing;

    let mut stream = stream::iter(files_to_import)
        .map(move |file_path| {
            let semaphore = semaphore.clone();
            let processing = processing.clone();

            async move {
                let _permit = semaphore.acquire().await.unwrap();

                if !file_path.exists() {
                    update_job_progress(
                        false,
                        Some(&format!("Missing file: {}", file_path.display())),
                    );
                    return;
                }

                let media_id = process_media_file(&file_path, &processing).await;

                if media_id.is_none() {
                    update_job_progress(
                        false,
                        Some(&format!("Failed to process: {}", file_path.display())),
                    );
                    return;
                }

                if delete_after_import {
                    if let Err(e) = tokio::fs::remove_file(&file_path).await {
                        update_job_progress(
                            false,
                            Some(&format!("Failed to delete {}: {}", file_path.display(), e)),
                        );
                        return;
                    }
                }

                update_job_progress(true, None);
            }
        })
        .buffer_unordered(effective_concurrency);

    while (stream.next().await).is_some() {}

    finalize_job_success();
}

pub async fn start_webdav_import_job(config: Arc<Config>, pool: DbPool) {
    if !config.webdav.enabled {
        info!("WebDAV import job disabled");
        return;
    }

    let poll_interval =
        std::time::Duration::from_secs(config.webdav.processing.poll_interval_seconds);

    info!(
        "Starting WebDAV import job: polling every {}s, root={}",
        config.webdav.processing.poll_interval_seconds,
        WEBDAV_DIR.display()
    );

    loop {
        run_webdav_import_cycle(&config, &pool).await;
        tokio::time::sleep(poll_interval).await;
    }
}

async fn run_webdav_import_cycle(config: &Config, pool: &DbPool) {
    if !WEBDAV_DIR.exists() {
        warn!(
            "WebDAV root directory missing, skipping import cycle: {}",
            WEBDAV_DIR.display()
        );
        return;
    }

    let Ok(entries) = std::fs::read_dir(&*WEBDAV_DIR) else {
        error!(
            "Failed to read WebDAV root directory: {}",
            WEBDAV_DIR.display()
        );
        return;
    };

    let semaphore = Arc::new(Semaphore::new(
        config.webdav.processing.max_concurrent_processing,
    ));

    let mut user_dir_count = 0usize;
    let mut skipped_user_dirs = 0usize;
    let mut queued_files = 0usize;
    let mut tasks: JoinSet<()> = JoinSet::new();

    for entry in entries.filter_map(|e| e.ok()) {
        let user_dir = entry.path();
        if !user_dir.is_dir() {
            continue;
        }

        let username = match user_dir.file_name().and_then(|n| n.to_str()) {
            Some(name) if !name.starts_with('.') => name.to_string(),
            _ => continue,
        };

        user_dir_count += 1;

        let user_id = match lookup_user_id(pool, &username) {
            Some(id) => id,
            None => {
                warn!("WebDAV directory for unknown user: {}", username);
                skipped_user_dirs += 1;
                continue;
            }
        };

        let stable_age = config.webdav.processing.stable_file_age_seconds;
        let files = collect_stable_webdav_files(&user_dir, stable_age);

        if files.is_empty() {
            debug!(
                "WebDAV import: no stable files for user {} (id={})",
                username, user_id
            );
            continue;
        }

        queued_files += files.len();
        info!(
            "WebDAV import: found {} stable files for user {} (id={})",
            files.len(),
            username,
            user_id
        );

        for file_path in files {
            let semaphore = semaphore.clone();
            let config = config.clone();
            let pool = pool.clone();
            let user_dir = user_dir.clone();

            tasks.spawn(async move {
                let _permit = semaphore.acquire().await.unwrap();

                process_webdav_file(&file_path, user_id, &user_dir, &config, &pool).await;
            });
        }
    }

    while tasks.join_next().await.is_some() {}

    if user_dir_count == 0 {
        debug!(
            "WebDAV import: no user directories found under {}",
            WEBDAV_DIR.display()
        );
        return;
    }

    debug!(
        "WebDAV import cycle complete: users_scanned={}, queued_files={}, skipped_users={}",
        user_dir_count, queued_files, skipped_user_dirs
    );
}

async fn process_webdav_file(
    file_path: &Path,
    user_id: i64,
    user_dir: &Path,
    config: &Config,
    pool: &DbPool,
) {
    let filename = file_path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("unknown");

    info!("WebDAV processing: {} for user {}", filename, user_id);

    let processing_dir = user_dir.join(".processing");
    if let Err(e) = std::fs::create_dir_all(&processing_dir) {
        error!(
            "Failed to create processing dir {}: {}",
            processing_dir.display(),
            e
        );
        return;
    }

    let processing_path = processing_dir.join(filename);
    if let Err(e) = std::fs::rename(file_path, &processing_path) {
        error!(
            "Failed to move file to processing: {} ({})",
            file_path.display(),
            e
        );
        return;
    }

    debug!(
        "WebDAV file moved to processing: {}",
        processing_path.display()
    );

    let processing = MediaProcessingContext {
        user_id,
        thumbnails: config.thumbnails.clone(),
        reverse_geocoding: Some(config.reverse_geocoding.clone()),
        pool: pool.clone(),
    };
    let result = process_media_file(&processing_path, &processing).await;

    match result {
        Some(media_id) => {
            info!(
                "WebDAV import success: {} -> media_id={} (thumbnails + metadata generated)",
                filename, media_id
            );
            match tokio::fs::remove_file(&processing_path).await {
                Ok(()) => {
                    debug!(
                        "WebDAV cleaned up processed file: {}",
                        processing_path.display()
                    );
                }
                Err(e) => {
                    warn!("Failed to cleanup processed file: {}", e);
                }
            }
        }
        None => {
            error!("WebDAV import failed: {}", filename);
            move_to_failed(&processing_path, user_dir).await;
        }
    }
}

async fn move_to_failed(processing_path: &Path, user_dir: &Path) {
    let failed_dir = user_dir.join(".failed");
    if let Err(e) = std::fs::create_dir_all(&failed_dir) {
        error!(
            "Failed to create failed dir {}: {}",
            failed_dir.display(),
            e
        );
        return;
    }

    let filename = processing_path.file_name().unwrap_or_default();

    let failed_path = failed_dir.join(filename);
    let error_sidecar = failed_dir.join(format!("{}.error.txt", filename.to_string_lossy()));

    if let Err(e) = std::fs::rename(processing_path, &failed_path) {
        error!(
            "Failed to move file to failed dir {}: {}",
            failed_path.display(),
            e
        );
        return;
    }

    debug!("WebDAV moved failed file to {}", failed_path.display());

    let error_content = format!(
        "Import failed at: {}\nOriginal path: {}",
        chrono::Utc::now().to_rfc3339(),
        processing_path.display()
    );

    match std::fs::write(&error_sidecar, error_content) {
        Ok(()) => {
            debug!("WebDAV wrote error sidecar: {}", error_sidecar.display());
        }
        Err(e) => {
            warn!("Failed to write error sidecar: {}", e);
        }
    }
}

fn lookup_user_id(pool: &DbPool, username: &str) -> Option<i64> {
    let conn = pool.get().ok()?;
    fetch_one(
        &conn,
        "SELECT id FROM users WHERE username = ? AND is_active = 1",
        &[&username],
        |row| row.get(0),
    )
    .ok()
    .flatten()
}

fn collect_stable_webdav_files(dir: &Path, stable_age_seconds: u64) -> Vec<PathBuf> {
    let mut files = Vec::new();
    let now = std::time::SystemTime::now();
    let min_age = std::time::Duration::from_secs(stable_age_seconds);

    collect_stable_files_recursive(dir, &mut files, now, min_age);
    files
}

fn collect_stable_files_recursive(
    dir: &Path,
    files: &mut Vec<PathBuf>,
    now: std::time::SystemTime,
    min_age: std::time::Duration,
) {
    let Ok(entries) = std::fs::read_dir(dir) else {
        return;
    };

    for entry in entries.filter_map(|e| e.ok()) {
        let path = entry.path();
        let name = path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or_default();

        if name.starts_with('.') {
            continue;
        }

        if path.is_dir() {
            collect_stable_files_recursive(&path, files, now, min_age);
        } else if path.is_file() {
            if let Ok(metadata) = path.metadata() {
                if let Ok(modified) = metadata.modified() {
                    if let Ok(age) = now.duration_since(modified) {
                        let supported = is_supported_extension(&path);
                        if age >= min_age && supported {
                            debug!("WebDAV file stable: {}", path.display());
                            files.push(path);
                        } else if age < min_age {
                            debug!(
                                "WebDAV file not stable yet: {} (age {}s < {}s)",
                                path.display(),
                                age.as_secs(),
                                min_age.as_secs()
                            );
                        } else if !supported {
                            debug!(
                                "WebDAV file skipped (unsupported extension): {}",
                                path.display()
                            );
                        }
                    }
                }
            }
        }
    }
}

fn is_supported_extension(path: &Path) -> bool {
    path.extension()
        .and_then(|e| e.to_str())
        .map(|e| format!(".{}", e.to_lowercase()))
        .map(|ext| SUPPORTED_EXTENSIONS.contains(ext.as_str()))
        .unwrap_or(false)
}
